### PR TITLE
Add support for timed ILS function blocking

### PR DESCRIPTION
**Pull request by rajaro**
Allows ILS driver functions to be blocked during certain times. Added example configuration to Demo.ini

### CODE REVIEW

File: module/VuFind/src/VuFind/ILS/Connection.php [hunk 1]
```diff
@@ -1325,6 +1328,96 @@ public function callIlsWithFailover($methodName, $params)
...
 
+    /**
+     * Get timed blocks for a method from driver configuration
+     *
+     * @param string $methodName Method to check
+     *
+     * @return array Array with keys 'start', 'end', 'recurring'
+     *               or empty array if no blocks are found
+     */
+    public function getMethodTimedBlocks(string $methodName): array
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Connection.php]: Does this need to be public? Would protected be better?

File: module/VuFind/src/VuFind/ILS/Connection.php [hunk 2]
```diff
@@ -1325,6 +1328,96 @@ public function callIlsWithFailover($methodName, $params)
...
+
+    /**
+     * Check whether a method is currently blocked in TimedBlocks section of
+     * driver configuration
+     *
+     * @param string $methodName Method to check
+     *
+     * @return bool
+     */
+    public function isMethodBlocked(string $methodName): bool
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Connection.php]: Same question: does this need to be public? (Admittedly, making it public would make it easier to write a unit test, which I think would be a great idea due to  all the variations and edge cases -- but if testing is the only reason, we could cheat with the ReflectionTrait, so it's not that big of a deal).

File: config/vufind/Demo.ini [hunk 3]
```diff
@@ -236,6 +236,16 @@ extraHoldFields = "comments:requestGroup:pickUpLocation:proxiedUsers:startDate:r
...
 
+; Block ILS methods during given times. Format is "method:setting = value"
+; method   - Method in ILS driver
+; settting - Valid settings are startDate, endDate, recurringStart, recurringEnd
+; value    - Date for startDate/endDate, time for recurringStart/recurringEnd
+;[TimedBlocks]
+;Renewals:startDate = 2025-07-01
+;Renewals:endDate = 2025-08-01
+;Renewals:recurringStart = 01:00
+;Renewals:recurringEnd = 03:00
```
Comment (demiankatz) [inline config/vufind/Demo.ini]: If you formatted this like:

```suggestion
;Renewals[startDate] = 2025-07-01
;Renewals[endDate] = 2025-08-01
;Renewals[recurringStart] = 01:00
;Renewals[recurringEnd] = 03:00
```

...I think you could simplify your config parsing a bit, since you'd have a formatted array right out of the .ini file without needing to split keys and rearrange data.

File: config/vufind/Demo.ini [hunk 4]
```diff
@@ -236,6 +236,16 @@ extraHoldFields = "comments:requestGroup:pickUpLocation:proxiedUsers:startDate:r
 ; Should we include a link to a fake OPAC?
 fakeOpacLink = false
 
+; Block ILS methods during given times. Format is "method:setting = value"
+; method   - Method in ILS driver
+; settting - Valid settings are startDate, endDate, recurringStart, recurringEnd
+; value    - Date for startDate/endDate, time for recurringStart/recurringEnd
```
Comment (demiankatz) [inline config/vufind/Demo.ini]: Maybe worth commenting on preferred date/time formats here (whether that's as simple as saying "compatible with strtotime" or a little more complex is up to you).

File: module/VuFind/src/VuFind/ILS/Connection.php [hunk 5]
```diff
@@ -1325,6 +1328,96 @@ public function callIlsWithFailover($methodName, $params)
...
+            [$method, $setting] = explode(':', $key, 2);
+            $methodBlocks[$method][$setting] = $value;
+        }
+        if (!isset($methodBlocks[$methodName])) {
+            return [];
+        }
+        $methodBlock = $methodBlocks[$methodName];
+        $startDate = strtotime($methodBlock['startDate'] ?? '');
+        $endDate = $methodBlock['endDate'] ?? '';
+        $noEndHours = empty(explode(' ', $endDate, 2)[1]);
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Connection.php]: This could influence the comments in the .ini file -- there are strtotime-compatible strings that contain spaces (e.g. "-2 days") but don't conform to your expectations here, so we may need to be more specific about what is allowed.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @rajaro! See below for some questions and comments. I definitely think some test coverage would be valuable to add before merging, as would be support for some real ILS drivers. (In some cases, I think this could be as simple as adding a section to the relevant .ini files, though in others, code changes will also be needed).

File: module/VuFind/src/VuFind/ILS/Connection.php (hunk 2 above)
Comment (rajaro) [reply module/VuFind/src/VuFind/ILS/Connection.php]: If these are public, they could be used to get the blocked times to display that information to the user somewhere in the UI

Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Connection.php]: Ahh, of course, that makes sense. I still think perhaps `getMethodTimedBlocks` would be better protected, since the information it returns feels like "internals" to me -- but I definitely agree that isMethodBlocked makes sense as part of the public interface.

Comment (rajaro) [reply module/VuFind/src/VuFind/ILS/Connection.php]: I think there are situations where `getMethodTimedBlocks` would be relevant to the user as well, as it might be useful to know when a certain functionality can be expected to work again

Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Connection.php]: True. But there's a lot of processing required to figure that out due to the number of possible settings. I wonder if having a public `getMethodTimedBlockEndDateTime` method (or something like that) would provide the useful external information in an easier-to-consume fashion. Or, alternatively, maybe we need a view helper (or other utility class) that can transform all of the settings in useful ways, so consumers of the method don't have to reinvent complicated logic.

File: config/vufind/Demo.ini (hunk 3 above)
Comment (rajaro) [reply config/vufind/Demo.ini]: Now simplified.

File: config/vufind/Demo.ini (hunk 4 above)
Comment (rajaro) [reply config/vufind/Demo.ini]: I went with instructing to use YYYY-MM-DD and HH:MM. Is that enough? `strtotime` will catch other formats as well, but as you pointed out in the other comment, something like "-2 days" might cause a problem.

Comment (demiankatz) [reply config/vufind/Demo.ini]: Yes, that makes sense to me!

File: config/vufind/Demo.ini [hunk 6]
```diff
@@ -236,6 +236,16 @@ extraHoldFields = "comments:requestGroup:pickUpLocation:proxiedUsers:startDate:r
 ; Should we include a link to a fake OPAC?
 fakeOpacLink = false
 
+; Block ILS methods during given times. Format is "method[setting] = value"
+; method  - Method in ILS driver
+; setting - Valid settings are startDate, endDate, recurringStart, recurringEnd
+; value   - Date (YYYY-MM-DD) for startDate/endDate, time (HH:MM) for recurringStart/recurringEnd
```
Comment (demiankatz) [inline config/vufind/Demo.ini]: Could this be `YYYY-MM-DD or YYYY-MM-DD HH:MM` since you could theoretically include a time in the start and end dates?

File: config/vufind/KohaRest.ini [hunk 7]
```diff
@@ -228,3 +228,13 @@ enabled = true
 ;log_function_result[createProfileArray] = true
 ; Log results createPatronArray into debug log file
 ;log_function_result[createPatronArray] = true
+
+; Block ILS methods during given times. Format is "method[setting] = value"
+; method  - Method in ILS driver
+; setting - Valid settings are startDate, endDate, recurringStart, recurringEnd
+; value   - Date (YYYY-MM-DD) for startDate/endDate, time (HH:MM) for recurringStart/recurringEnd
```
Comment (demiankatz) [inline config/vufind/KohaRest.ini]: If this changes above, we need to remember to synchronize changes here.

File: config/vufind/SierraRest.ini [hunk 8]
```diff
@@ -294,3 +294,13 @@ identifier = "@"
 ;log_function_result[createProfileArray] = true
 ; Log results createPatronArray into debug log file
 ;log_function_result[createPatronArray] = true
+
+; Block ILS methods during given times. Format is "method[setting] = value"
+; method  - Method in ILS driver
+; setting - Valid settings are startDate, endDate, recurringStart, recurringEnd
+; value   - Date (YYYY-MM-DD) for startDate/endDate, time (HH:MM) for recurringStart/recurringEnd
```
Comment (demiankatz) [inline config/vufind/SierraRest.ini]: ...and here.

File: module/VuFind/src/VuFind/ILS/Connection.php [hunk 9]
```diff
@@ -1325,6 +1328,89 @@ public function callIlsWithFailover($methodName, $params)
...
+
+        if (!isset($functionConfig[$methodName])) {
+            return [];
+        }
+        $methodBlock = $functionConfig[$methodName];
+        $startDate = strtotime($methodBlock['startDate'] ?? '');
+        $endDate = $methodBlock['endDate'] ?? '';
+        $noEndHours = empty(explode(' ', $endDate, 2)[1]);
+        if ($endDate && $noEndHours) {
+            $endDate .= ' 23.59.59';
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Connection.php]: If we recommend HH:MM in the config, should this be ` 23:59.59` for consistency?

File: config/vufind/Demo.ini [hunk 10]
```diff
@@ -236,6 +236,16 @@ extraHoldFields = "comments:requestGroup:pickUpLocation:proxiedUsers:startDate:r
 ; Should we include a link to a fake OPAC?
 fakeOpacLink = false
 
+; Block ILS methods during given times. Format is "method[setting] = value"
+; method  - Method in ILS driver
+; setting - Valid settings are startDate, endDate, recurringStart, recurringEnd
```
Comment (demiankatz) [inline config/vufind/Demo.ini]: Maybe it's worth adding a little more explanation, like:

startDate and endDate specify fixed start and end dates (with optional times) to make the method unavailable. You do not need to specify both -- you can use either on its own for an open-ended range.

recurringStart and recurringEnd specify a time window that will block the method every day. If recurringEnd is lower than recurringStart, the driver will assume a time window crossing from one day into the next day. When setting a recurring window, both recurringStart and recurringEnd MUST be set.

If startDate or endDate is specified, the recurring settings will be ignored; you cannot combine these two types of settings.

File: module/VuFind/tests/unit-tests/src/VuFindTest/ILS/ConnectionTest.php [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+    /**
+     * Standard setup method.
+     *
+     * @return void
+     */
+    public function setUp(): void
+    {
+        $config = new Config(['driver' => 'Demo']);
+        $driverManager = $this->getMockBuilder(\VuFind\ILS\Driver\PluginManager::class)
+            ->disableOriginalConstructor()->getMock();
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/ILS/ConnectionTest.php]: You can simplify to:

```suggestion
        $driverManager = $this->createMock(\VuFind\ILS\Driver\PluginManager::class);
```

File: module/VuFind/tests/unit-tests/src/VuFindTest/ILS/ConnectionTest.php [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+        $driverManager->method('has')->willReturn('Demo');
+        $configReader = $this->getMockBuilder(\VuFind\Config\PluginManager::class)
+            ->disableOriginalConstructor()->getMock();
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/ILS/ConnectionTest.php]: This can also be simplified:

```suggestion
        $configReader = $this->createMock(\VuFind\Config\PluginManager::class);
```

File: module/VuFind/tests/unit-tests/src/VuFindTest/ILS/ConnectionTest.php [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+     * Set TimedBlocks driver configuration
+     *
+     * @param array $timedBlocks timed blocks as defined in driver.ini
+     *
+     * @return void
+     */
+    public function setTimedBlocks(array $timedBlocks): void
+    {
+        $driver = $this->getMockBuilder(\VuFind\ILS\Driver\Demo::class)
+            ->disableOriginalConstructor()->getMock();
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/ILS/ConnectionTest.php]: And this:

```suggestion
        $driver = $this->createMock(\VuFind\ILS\Driver\Demo::class);
```

File: module/VuFind/tests/unit-tests/src/VuFindTest/ILS/ConnectionTest.php [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+            $config,
+            $driverManager,
+            $configReader
+        );
+    }
+
+    /**
+     * Set TimedBlocks driver configuration
+     *
+     * @param array $timedBlocks timed blocks as defined in driver.ini
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/ILS/ConnectionTest.php]: Should we use the actual filename here?

```suggestion
     * @param array $timedBlocks timed blocks as defined in Demo.ini
```

File: module/VuFind/tests/unit-tests/src/VuFindTest/ILS/ConnectionTest.php [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+
+        $this->connection->setDriver($driver);
+    }
+
+    /**
+     * Test that methods are blocked with only startDate defined and the block starts now
+     *
+     * @return void
+     */
+    public function testIsMethodBlockedWithStartDate()
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/ILS/ConnectionTest.php]: Have you ever worked with a PHPUnit dataProvider before? Since all of these tests are using essentially the same logic, I think it might make sense to refactor them to a dataProvider method plus a test with a signature like:

```
public function testIsMethodBlocked(array $timedBlocks, bool $expectedResult)
```

If you need an example of how this works, just search the code for the `@dataProvider` annotation -- or let me know if you have any questions!

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @rajaro!

See below for some suggestions about further improving the documentation and potentially simplifying the test. The test changes are not essential if you don't think it's worth the time to make the changes (we could do it in the future too), but I thought they were worth suggesting in case you found the ideas helpful.

File: module/VuFind/src/VuFind/ILS/Connection.php (hunk 2 above)
Comment (rajaro) [reply module/VuFind/src/VuFind/ILS/Connection.php]: I think it makes sense to have a view helper that fetches the relevant information. `getMethodBlocks` is now protected and the view helper provides necessary information (both start and end dates).

File: module/VuFind/tests/unit-tests/src/VuFindTest/ILS/ConnectionTest.php (hunk 15 above)
Comment (rajaro) [reply module/VuFind/tests/unit-tests/src/VuFindTest/ILS/ConnectionTest.php]: Yes, and this was a pretty obvious place to use it. Changed.

File: module/VuFind/tests/unit-tests/src/VuFindTest/ILS/ConnectionTest.php [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+     * @param bool  $expectedResult The expected result
+     *
+     * @dataProvider isMethodBlockedProvider
+     *
+     * @return void
+     */
+    public function testIsMethodBlocked(array $timedBlocks, bool $expectedResult): void
+    {
+        $this->setTimedBlocks($timedBlocks);
+        $this->assertEquals($this->connection->isMethodBlocked('Renewals'), $expectedResult);
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/ILS/ConnectionTest.php]: The expected value should come first to ensure appropriate messages in case of failure:

```suggestion
        $this->assertEquals($expectedResult, $this->connection->isMethodBlocked('Renewals'));
```

File: module/VuFind/src/VuFind/View/Helper/Root/MethodTimedBlocks.php [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+class MethodTimedBlocks extends \Laminas\View\Helper\AbstractHelper
+{
+    /**
+     * Get TimedBlocks settings from driver configuration
+     *
+     * @param string $methodName Method to check
+     *
+     * @return array
+     */
+    public function __invoke(string $methodName): array
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/View/Helper/Root/MethodTimedBlocks.php]: This isn't exactly what I had in mind, but you are the one who best understands the use case, so maybe that's okay.

I was thinking that, if we kept getMethodTimedBlocks public, the view helper would turn that data into a human-readable message, like "This feature is unavailable until X."

It instead seems that this is just reading the raw configuration and passing it through without a lot of processing; I'm not sure if that makes it possible to construct general-purpose messages, especially when dealing with some of the nuances related to recurring time windows.

I guess I have a couple of general concerns:

1.) This still seems to be creating parallel logic -- i.e. we're reading and processing the configs in two different places rather than sharing logic. I'd feel happier if one piece of logic fed into the other somehow.

2.) This isn't actually used anywhere -- it might be easier to understand and critique if we had an example usage. Would it make sense to introduce something somewhere? If not, maybe we should at least build a unit test to demonstrate the code in action (and perhaps document the existence of the helper in the .ini files so people know they can leverage it).

Please let me know what you think! We could also just drop the view helper idea until we actually need it, if it's too much work/nuisance to refine this further right now. I want to help improve the design, but I don't want to waste your time on theoretical concerns if you don't actually need this. :-)

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @rajaro; see below for my latest comments!

File: module/VuFind/src/VuFind/View/Helper/Root/MethodTimedBlocks.php (hunk 17 above)
Comment (rajaro) [reply module/VuFind/src/VuFind/View/Helper/Root/MethodTimedBlocks.php]: I agree that the parallel logic should be removed and the logic should stay in one place. But with regards to the message, I think there's some issues. Only having a general message like "This feature is unavailable until X." feels a little too vague to me. What should be displayed if no end date is defined? And if we're on the checked out page, and there is a message saying "This feature is unavailable until X", that might leave too much for interpretation. I think that the message should clearly say what feature is unavailable. 

The original idea was to simply make it possible to fetch the blocked times and build the message string in the template due to different needs and situations. Having a view helper do this does make it cleaner though. With a view helper, I'm thinking there could be some different solutions:

1. Have the view helper process the dates and return a translated string based on the configuration (for example `method_blocked_during`, `method_blocked_until`, `method_blocked_recurring`). This could also support different translations for different functions with something like `method_blocked_during_Renewals`, with a fallback to `method_blocked_during`
2. Have the helper return a general message like "This feature is unavailable until X." and if some other specification is needed, it could be just added in the template

Thoughts?

And to your points:
1.) Agreed. I will remove the duplicate logic.
2.) Sure, when we've settled on the solution, I'll at least write up a test and document the view helper in the .ini file.

Comment (demiankatz) [reply module/VuFind/src/VuFind/View/Helper/Root/MethodTimedBlocks.php]: I like the idea of the view helper having one or more message templates for various scenarios, and maybe accepting overrides where appropriate. So, for example, strings like:

%%service%% is unavailable until %%enddate%%

%%service%% is currently unavailable

Then %%service%% could default to "This feature" unless a more specific descriptor is passed to the helper. And maybe the helper could also accept overrides to the templates if you want to more heavily customize things. You would know better than me exactly how many templates are needed to cover all use cases and what placeholders should be made available... but it seems like a solvable problem. :-)

Does that make any sense? Is that more or less consistent with your own thoughts, or did you have something substantially different in mind?

Comment (rajaro) [reply module/VuFind/src/VuFind/View/Helper/Root/MethodTimedBlocks.php]: Makes sense. Now the view helper returns those message templates when appropriate. I think that should be enough, overriding translations is probably not necessary. The important information is the disabled feature and the end of the block.

File: languages/en.ini [hunk 18]
```diff
@@ -1515,6 +1517,7 @@ The system is currently unavailable due to system maintenance = "The system is c
 Theme = "Theme"
 Thesis = "Thesis"
 This email was sent from = "This email was sent from"
+This feature = "This feature"
```
Comment (demiankatz) [inline languages/en.ini]: For context, maybe you could use as a key here, `default_service_description` instead of `This feature`.

File: languages/en.ini [hunk 19]
```diff
@@ -862,6 +862,8 @@ MeSH Terms = "MeSH Terms"
 Message = "Message"
 Message From Sender = "Message From Sender"
 Metadata Prefix = "Metadata Prefix"
+method_blocked = "%%service%% is currently unavailable"
+method_blocked_until = "%%service%% is unavailable until %%end%%"
```
Comment (demiankatz) [inline languages/en.ini]: And maybe `service_blocked` / `service_blocked_until` would be more meaningful to the translators than `method`, which refers more to the underlying implementation details than what is actually happening from a high-level perspective.

File: module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/MethodTimedBlocksTest.php [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+     * Get view helpers needed by test.
+     *
+     * @param array $timedBlocks Timed blocks
+     * @param bool  $blocked     Is the method blocked
+     *
+     * @return array
+     */
+    protected function getViewHelpers(array $timedBlocks, bool $blocked)
+    {
+        $getTranslation = function ($str, $tokens = []) {
```
Comment (demiankatz) [inline module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/MethodTimedBlocksTest.php]: This mock translation view helper logic feels like it might be useful to reuse. I wonder if it's worth refactoring into https://github.com/vufind-org/vufind/blob/dev/module/VuFind/src/VuFindTest/Feature/TranslatorTrait.php (and I also wonder if we already have logic like this in other places). I'm happy to accept this as-is, but if you have a couple of minutes, it might be worth doing a skim to see if we can unify any of this logic to simplify our tests.

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @rajaro! See below for a few more comments -- the suggested naming improvements are probably the most important; you can take or leave my ideas about the tests. :-)

File: module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/MethodTimedBlocksTest.php (hunk 20 above)
Comment (rajaro) [reply module/VuFind/tests/unit-tests/src/VuFindTest/View/Helper/Root/MethodTimedBlocksTest.php]: Just using `TranslatorTrait` works and simplifies the test nicely. The same logic was used in `FlashmessagesTest.php` and I refactored that to use `TranslatorTrait` as well.

**Review by demiankatz — APPROVED** [review]
Thanks, @rajaro -- this looks good to me now, and all tests are passing. I'm going to leave this open a little longer to give others an opportunity to offer feedback, in case I've missed anything... but I'll merge it soon! I've tagged it for inclusion in release 11.0 so it won't get forgotten.

File: module/VuFind/src/VuFind/ILS/Connection.php [hunk 21]
```diff
@@ -1325,6 +1328,89 @@ public function callIlsWithFailover($methodName, $params)
...
+     * Get timed blocks for a method from driver configuration
+     *
+     * @param string $methodName Method to check
+     *
+     * @return array Array with keys 'start', 'end', 'recurring'
+     *               or empty array if no blocks are found
+     */
+    public function getMethodTimedBlocks(string $methodName): array
+    {
+        $functionConfig = $this->checkCapability('getConfig', ['TimedBlocks'])
```
Comment (EreMaijala) [inline module/VuFind/src/VuFind/ILS/Connection.php]: These won't work with MultiBackend driver without a parameter containing a source.

File: config/vufind/Demo.ini [hunk 22]
```diff
@@ -236,6 +236,23 @@ extraHoldFields = "comments:requestGroup:pickUpLocation:proxiedUsers:startDate:r
...
+; value   - Date (YYYY-MM-DD or YYYY-MM-DD HH:MM) for startDate/endDate, time (HH:MM) for recurringStart/recurringEnd
+; startDate and endDate specify fixed start and end dates (with optional times) to make the method unavailable.
+; You do not need to specify both -- you can use either on its own for an open-ended range.
+; recurringStart and recurringEnd specify a time window that will block the method every day.
+; If recurringEnd is lower than recurringStart, the driver will assume a time window crossing from one day into the next day.
+; When setting a recurring window, both recurringStart and recurringEnd MUST be set.
+; If startDate or endDate is specified, the recurring settings will be ignored; you cannot combine these two types of settings.
+; Timed blocks can be displayed using MethodTimedBlocks view helper.
+;[TimedBlocks]
```
Comment (EreMaijala) [inline config/vufind/Demo.ini]: Can we keep the section name uncommented? It's really easy to miss that you need to uncomment it too.

File: module/VuFind/src/VuFind/ILS/Connection.php [hunk 23]
```diff
@@ -1325,6 +1328,89 @@ public function callIlsWithFailover($methodName, $params)
...
+
+        if (!isset($functionConfig[$methodName])) {
+            return [];
+        }
+        $methodBlock = $functionConfig[$methodName];
+        $startDate = strtotime($methodBlock['startDate'] ?? '');
+        $endDate = $methodBlock['endDate'] ?? '';
+        $noEndHours = empty(explode(' ', $endDate, 2)[1]);
+        if ($endDate && $noEndHours) {
+            $endDate .= ' 23:59.59';
```
Comment (EreMaijala) [inline module/VuFind/src/VuFind/ILS/Connection.php]: Why period instead of colon?

File: module/VuFind/src/VuFind/ILS/Connection.php [hunk 24]
```diff
@@ -1325,6 +1328,89 @@ public function callIlsWithFailover($methodName, $params)
...
+                'recurring' => false,
+            ];
+        }
+        if (
+            ($startTime = strtotime($methodBlock['recurringStart'] ?? ''))
+            && ($endTime = strtotime($methodBlock['recurringEnd'] ?? ''))
+        ) {
+            return [
+                'start' => $startTime,
+                'end' => $endTime,
```
Comment (EreMaijala) [inline module/VuFind/src/VuFind/ILS/Connection.php]: How about modifying end time here if it's sooner than start time? I think it would make it clearer since the consumer of this information wouldn't need to check for this special case. If you'd parse the config with `new \DateTime(<setting>)` it would be easy to just add a day to it.

**Review by EreMaijala — CHANGES_REQUESTED** [review]
- Shouldn't the MethodTimedBlocks helper be used by default? Now there's no indication that renewals are blocked. It just looks like none of the items are renewable.
- I think it would be useful to have a list of common functions that could be blocked since they're not all quite obvious.
- What's also bugging me a bit is that the configuration doesn't allow for more than one interval. I don't think it's a too far-fetched case where two or more intervals would be useful. If the configuration was more like this, it would be more flexible:
```
[TimedBlocks]
Renewals[] = "2025-07-01/2025-08-01"
Renewals[] = "23:00/01:00"

```
Anything with a date would be non-recurring by nature. Using a slash as the separator would make this a subset of ISO 8601 and allow for extension to full ISO 8601 duration support later if needed.

File: module/VuFind/src/VuFind/ILS/Connection.php (hunk 21 above)
Comment (rajaro) [reply module/VuFind/src/VuFind/ILS/Connection.php]: I'm missing something here. It seems to work with MultiBackend, as `MultiBackend::getConfig` parses the source from `$this->ilsAuth->getStoredCatalogCredentials()`. Or is the problem when the user is not logged in?

File: module/VuFind/src/VuFind/ILS/Connection.php (hunk 23 above)
Comment (demiankatz) [reply module/VuFind/src/VuFind/ILS/Connection.php]: `23:59.59` is still used in a couple of places in MethodTimedBlocksTest.php -- is this intentional?

Comment (rajaro) [reply module/VuFind/src/VuFind/ILS/Connection.php]: Fixed.

**Comment by demiankatz** [comment]
@rajaro and @EreMaijala: there are a few unresolved conversation threads here. Have these been addressed in recent code changes, or is there still work in progress? Just trying to figure out whether you need me to take any action from my end right now.

File: module/VuFind/src/VuFind/ILS/Connection.php (hunk 21 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/ILS/Connection.php]: Exactly, when the user isn't logged in. Also, `getStoredCatalogCredentials()` returns the currently active information, and if the user is, for instance, trying to add a library card for another ILS, it should be able to retrieve the configuration for that ILS instead of the active one.

File: module/VuFind/src/VuFind/View/Helper/Root/MethodTimedBlocks.php [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+{
+    /**
+     * Returns a display string of timed blocks from driver configuration
+     *
+     * @param string $methodName        Method to check
+     * @param string $methodDisplayName Method display name translation key (Defaults to "This feature")
+     *
+     * @return string
+     */
+    public function __invoke(
```
Comment (EreMaijala) [inline module/VuFind/src/VuFind/View/Helper/Root/MethodTimedBlocks.php:52]: Also this method should be able to take params (e.g. user) that could be passed on to MultiBackend driver to select the correct ILS.

**Review by EreMaijala — CHANGES_REQUESTED** [review]

File: module/VuFind/src/VuFind/ILS/Connection.php [hunk 26]
```diff
@@ -1341,6 +1344,92 @@ public function callIlsWithFailover($methodName, $params)
...
+                }
+                $endDate = $end ? new \DateTime($end) : null;
+                $blocks[] = [
+                    'start' => $startDate,
+                    'end' => $endDate,
+                    'recurring' => false,
+                ];
+            } else {
+                $startTime = new \DateTime($start);
+                $endTime = new \DateTime($end);
```
Comment (EreMaijala) [inline module/VuFind/src/VuFind/ILS/Connection.php]: Shouldn't these check for empty $start or $end and create the objects only for non-empty? Otherwise the check right below doesn't make sense.

Comment (EreMaijala) [reply module/VuFind/src/VuFind/ILS/Connection.php]: Also needs tests to ensure that mistakes like "/10:00:00" and "10:00:00/" are handled properly.

**Review by EreMaijala — CHANGES_REQUESTED** [review]

**Review by EreMaijala — APPROVED** [review]
Looks good to me now!

**Review by demiankatz — CHANGES_REQUESTED** [review]
Thanks, @rajaro. Unfortunately, some tests are failing in my environment:

```
There were 2 failures:

1) VuFindTest\Mink\AccountActionsTest::testRecoverILSPassword with data set #0 (false, false)
Element not found: .modal-body .recover-account-link index 0
Failed asserting that null is of type object.

/usr/local/vufind/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:569
/usr/local/vufind/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:663
/usr/local/vufind/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/AccountActionsTest.php:704

2) VuFindTest\Mink\AccountActionsTest::testRecoverILSPassword with data set #1 (false, true)
Element not found: .modal-body .recover-account-link index 0
Failed asserting that null is of type object.

/usr/local/vufind/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:569
/usr/local/vufind/module/VuFind/src/VuFindTest/Integration/MinkTestCase.php:663
/usr/local/vufind/module/VuFind/tests/integration-tests/src/VuFindTest/Mink/AccountActionsTest.php:704

FAILURES!
Tests: 18, Assertions: 442, Failures: 2.
```

I think the root cause is this error in my exception log:

```
2025-10-07T08:10:12-04:00 CRIT (2): TypeError : VuFind\ILS\Connection::getMethodBlock(): Argument #2 ($params) must be of type array, null given, called in /usr/local/vufind/module/VuFind/src/VuFind/ILS/Connection.php on line 364
```

**Comment by demiankatz** [comment]
@rajaro, I pushed up https://github.com/vufind-org/vufind/pull/4550/commits/b27a5247a8542569f5d502e47b4bacdd9b5af35c to fix the above issue. Still testing on my end...

**Review by demiankatz — APPROVED** [review]
My fix seems to have sorted things out. Just waiting for a final full test run to complete, and if no problems occur, I'll merge this.
